### ISSUE TITLE
Jedis should not throw an exception on close()

### ISSUE THREAD

**Issue by eric-wu** [xetorthio/jedis#992]
This is about Jedis version 2.7.2.

The class `Jedis` implements `java.io.Closeable`.  The Java doc for the `Closeable.close()` clearly states "If the stream is already closed then invoking this method has no effect.".  However, this is not true with `Jedis.close()`.  An IllegalStateException is wrapped into a JedisException and is thrown to the caller, which is a surprise.  I would suggest Jedis swallow the IllegalStateException thrown by GenericObjectPool instead of passing it to the caller.

**Comment by marcosnils** [comment]
@eric-wu just tried this code in 2.7.2 and I'm not getting any exception

```java
public class MainTest {
	
	public static void main(String[] args) {
		Jedis jedis = new Jedis();
		jedis.close();
		jedis.close();
		System.out.println("Done");
	}

}
```

Can you please share a snippet which reproduces the error?

Thx!.

**Comment by eric-wu** [comment]
Can reproduce with the following (which is close to the actual code):

```
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;

public class JedisCloseTest {

    public static void main(String[] args) {
        JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setMaxTotal(10);
        String host = "localhost";
        int port = 6379;
        int timeout = 10;
        try (JedisPool jedisPool = new JedisPool(poolConfig, host, port, timeout)) {
            Jedis jedis = jedisPool.getResource();
            jedis.close();
            jedis.close();
        }
    }
}

```

**Comment by eric-wu** [comment]
We are on Java 8 and the packages are:

redis.clients:jedis:2.7.2
org.apache.commons:commons-pool2:2.3

**Comment by marcosnils** [comment]
Hmmm now I see the error. Seems like the error comes from commons-pool when trying to return an already returned resource. 

The thing is that commons-pool also raises `IllegalStateException` when calling `returnObject` for other reasons (such as object not corresponding to the underlying pool). If we "ignore" the IllegalState exceptions when calling `close` that may lead to possible errors which will be very difficult to track. 

I'm wondering which would be the best approach here. @eric-wu any suggestions?
@HeartSaVioR, @xetorthio ?

**Comment by eric-wu** [comment]
"such as object not corresponding to the underlying pool" -- that is indeed a valid exception. Can it ever happen with how Jedis uses the object pool?

**Comment by marcosnils** [comment]
@eric-wu in jedis 2.7 and 2.6 it may happen if you use the deprecated functions `JedisPool.returnResource` or `JedisPool.returnBrokenResource`. Maybe we can include this change in Jedis 3.0 to avoid breaking current deprecated methods. 

What do you think?

**Comment by eric-wu** [comment]
I see.  Your plan sounds reasonable to me.

I can also think of other ways to work around the exception thrown by returnObject().  For example, Jedis can augment GenericObjectPool by adding a synchronized hash map to track the idle connections in the pool.  Or in a similar way each connection object can track its state (in-use, returned, etc.) with an enum field.  Just to throw in some ideas here.  Not sure if they make sense.

By the way, I appreciate the super quick response on this issue.

**Comment by marcosnils** [comment]
@eric-wu I believe I've found a better way to fix this. Please check #993

**Comment by eric-wu** [comment]
Nice! And thanks!

**Comment by HeartSaVioR** [comment]
Hmm... It's not an easy issue for me.

Closing twice can occur on 3.0 since users can call Jedis.close() twice, or use try-with-resource and Jedis.close() together. So we cannot defense wrong usages.

Way of resolving issue... Actually I'm with @eric-wu though it's not an easy way since I can't think about other ways. If we agree to go on, then we should check it can be possible.

**Comment by marcosnils** [comment]
Or.. we can be bit a bit more aggressive and catch `IllegalStateException` when `returnObject` or `invalidateObject` is called and include it in 2.6 and 2.7. After all `JedisPool.returnResource` and `JedisPool.returnBrokenResource` are **deprecated** and shall **NOT** be used.

If we get any issues from users in 2.6 or 2.7 using the deprecated methods we need to first advice them to stop using those functions.

**Comment by marcosnils** [comment]
@HeartSaVioR what do you think?

**Comment by eric-wu** [comment]
Just let you know that my original proposal, that Jedis swallows the IllegalStateException thrown by GenericObjectPool, is incorrect.  We implemented in our code to ignore the exception as a temporary workaround.  Then we saw occasional connection issues (see [BRIDGE-572](https://sagebionetworks.jira.com/projects/BRIDGE/issues/BRIDGE-572)).  We have to make sure that a connection is closed only once and then the problem is gone.
